dvc/project.py

```diff
@@ -0,0 +1,203 @@
+            self.stages[node].reproduce()
+
+        stage.reproduce()
+
+
+class Project(object):
+    DVC_DIR = '.dvc'
+
+    def __init__(self, root_dir):
+        self._stages = None
+        self._pipelines = None
+        self._G = None
+
+        self.root_dir = os.path.abspath(root_dir)
+        self.dvc_dir = os.path.join(self.root_dir, self.DVC_DIR)
+
+        self.scm = SCM(self.root_dir)
+        self.lock = Lock(self.dvc_dir)
+        self.cache = Cache(self.dvc_dir)
+        self.state = State(self.root_dir, self.dvc_dir)
+        self.config = Config(self.dvc_dir)
+        self.logger = Logger()
+        self.cloud = DataCloud(self.config._config)
+
+    @staticmethod
+    def init(root_dir):
+        root_dir = os.path.abspath(root_dir)
+        dvc_dir = os.path.join(root_dir, Project.DVC_DIR)
+        os.mkdir(dvc_dir)
+
+        config = Config.init(dvc_dir)
+        cache = Cache.init(dvc_dir)
+        state = State.init(root_dir, dvc_dir)
+        lock = Lock(dvc_dir)
+
+        scm = SCM(root_dir)
+        scm.ignore_list([cache.cache_dir,
+                         state.state_file,
+                         lock.lock_file])
+        scm.commit('DVC init')
+
+        return Project(root_dir)
+
+    def run(self, cmd, deps, deps_no_cache, outs, outs_no_cache, locked, fname, cwd):
+        cwd = os.path.abspath(cwd)
+        path = os.path.join(cwd, fname)
+        outputs = Output.loads_from(self, outs, use_cache=True, cwd=cwd)
+        outputs += Output.loads_from(self, outs_no_cache, use_cache=False, cwd=cwd)
+        deps = Dependency.loads_from(self, deps, use_cache=True, cwd=cwd)
+        deps += Dependency.loads_from(self, deps_no_cache, use_cache=False, cwd=cwd)
+
+        stage = Stage(project=self,
+                      path=path,
+                      cmd=cmd,
+                      cwd=cwd,
+                      outs=outputs,
+                      deps=deps,
+                      locked=locked)
+        stage.run()
+        stage.dump()
+
+    def reproduce(self, targets, recursive=False, force=False):
+        stages = nx.get_node_attributes(self.graph(), 'stage')
+        for target in targets:
+            node = os.path.relpath(os.path.abspath(target), self.root_dir)
+            if node not in stages:
+                raise StageNotFoundError(target)
+
+            if recursive:
+                for node in nx.dfs_postorder_nodes(self.graph(), node):
+                    stages[node].reproduce(force=force)
+
+            stages[node].reproduce(force=force)
+
+    def checkout(self):
+        for stage in self.stages():
+            stage.checkout()
+
+    def _used_cache(self):
+        clist = []
+        for stage in self.stages():
+            for entry in itertools.chain(stage.outs, stage.deps):
+                if not entry.use_cache:
+                    continue
+                if entry.cache not in clist:
+                    clist.append(entry.cache)
+        return clist
+
+    def gc(self):
```

Comment:
Probably this is not the best place for `gc` logic since a lot of GC strategies will be needed in the future: gc for branches for example.

Reply:
I couldn't find a better place for now, so i'll just leave it there. At least we have new internal design to follow =).